gobblin-compaction/src/main/java/gobblin/compaction/mapreduce/MRCompactor.java

```diff
@@ -294,39 +373,357 @@ MRCompactorJobPropCreator getJobPropCreator(String topic, Path topicInputDir, Pa
+          hadoopJob.killJob();
+        }
+      }
+    } finally {
+      try {
+        ExecutorsUtils.shutdownExecutorService(this.jobExecutor, Optional.of(LOG), 0, TimeUnit.NANOSECONDS);
+      } finally {
+        this.verifier.closeNow();
+      }
+    }
+  }
+
+  /**
+   * A subclass of {@link ThreadPoolExecutor} for running compaction jobs, and performs necessary steps
+   * after each compaction job finishes.
+   */
+  private class JobRunnerExecutor extends ThreadPoolExecutor {
+
+    public JobRunnerExecutor(int corePoolSize, int maximumPoolSize, long keepAliveTime, TimeUnit unit,
+        BlockingQueue<Runnable> workQueue) {
+      super(corePoolSize, maximumPoolSize, keepAliveTime, unit, workQueue);
+    }
+
+    /**
+     * When a compaction job for a {@link Dataset} finishes, if it successfully published the data (t == null
+     * && jobRunner.status() == {@link MRCompactorJobRunner.Status#COMMITTED}, or if it
+     * threw any {@link Throwable} (t != null), mark the {@link Dataset} as
+     * {@link Dataset.DatasetState#COMPACTION_COMPLETE}.
+     * If the job failed to publish the data because the input data was not complete, reduce the priority of
+     * the {@link Dataset}. A new compaction job will be submitted later with a lower priority.
+     */
+    @Override
+    protected void afterExecute(Runnable r, Throwable t) {
+      Preconditions.checkArgument(r instanceof MRCompactorJobRunner,
+          String.format("Runnable expected to be instance of %s, actual %s", MRCompactorJobRunner.class.getSimpleName(),
+              r.getClass().getSimpleName()));
+
+      MRCompactorJobRunner jobRunner = (MRCompactorJobRunner) r;
+      MRCompactor.this.jobRunnables.remove(jobRunner.getDataset());
+      if (t == null) {
+        if (jobRunner.status() == COMMITTED) {
+
+          // Compaction job of a dataset is successful.
+          LOG.info(String.format("Compacted data for %s successfully published", jobRunner.getDataset()));
+          jobRunner.getDataset().setState(COMPACTION_COMPLETE);
+        } else
+          if (jobRunner.getDataset().state() == GIVEN_UP && !MRCompactor.this.shouldPublishDataIfCannotVerifyCompl) {
+
+          // Compaction job of a dataset has aborted, and data completeness verification has given up.
+          // This dataset will not be compacted.
+          LOG.info(String.format("Dataset %s will not be compacted, since data completeness cannot be verified",
+              jobRunner.getDataset()));
+          jobRunner.getDataset().setState(COMPACTION_COMPLETE);
+        } else {
+
+          // Compaction job of a dataset has aborted because data completeness is not verified.
+          // Reduce priority and try again.
+          jobRunner.getDataset().reducePriority();
+        }
+      } else {
+
+        // Compaction job of a dataset has failed with a throwable.
+        jobRunner.getDataset().setState(COMPACTION_COMPLETE);
```

Comment:
For these kind of state transitions where multiple attributes need to be changed, it is better to have a helperMethod, e.g. completeWithError(Throwable) to make sure that the object is not in an inconsistent state.